### PR TITLE
hyper 0.13 + futures 0.3

### PR DESCRIPTION
**Pull request by robjtede**
Updates to hyper 0.13, futures 0.3 and many methods to async/await.

~NOTE: I've made many many terrible decisions in this PR so far in the name of getting it to actually compile and run. Given this bottleneck is lifted, it's possible to distrubute tasks out with the code as it is despite it being less-than-optimal. Things I broke are generally marked with TODOs.~ See comment.

### CODE REVIEW

**Comment by robjtede** [comment]
Most notably:
- I disabled the caching since it used the Shared structure since it's method and lifetime requirements are incompatible with the hyper service trait
- Calls to the cached structures in the Engine now require mut deref access (due to service trait) which just Arcs could not provide. Quick fix was to put std mutex in there but this is very not okay.

**Review by Enet4 — COMMENTED** [review]
> In the interests of keeping this PR as simple as possible (which it probably is), I suggest we make issues from these points and consider them known problems with the PR to be addressed as priority.

This is OK, considering that we're interested in splitting work load over fixing all problems at once. If these separate issues are filed now, we can refer to them in the TODOs.

Nevertheless, a bit of feedback over these changes follows. Nothing that should block this PR, as established.

**Review by paolobarbolini — APPROVED** [review]
I'm ok with keeping this simple. By merging this as is we allow for the rest of the work to be done in parallel.

File: src/interactors/crates.rs [hunk 1]
```diff
@@ -94,33 +103,41 @@ where
             _ => format!("{}/{}/{}", &lower_name[0..2], &lower_name[2..4], lower_name),
         };
 
-        let uri =
-            try_future_box!(format!("{}/master/{}", CRATES_INDEX_BASE_URI, path).parse::<Uri>());
+        let uri = format!("{}/master/{}", CRATES_INDEX_BASE_URI, path);
 
-        let request = Request::get(uri.clone()).body(Body::empty()).unwrap();
+        println!("analyze from uri {:?}", &uri);
```
[the remark this answers is not in the document]
Comment (robjtede) [reply src/interactors/crates.rs:107]: Good catch, this is debugging cruft and can be removed. But yes, having never worked with `slog` (vs `log`) it was strange to need access to the logging object at this point.

File: src/engine/mod.rs [hunk 2]
```diff
@@ -132,141 +140,150 @@ impl AnalyzeDependenciesOutcome {
...
+            None => Err(anyhow!(
+                "could not find crate release with version {}",
+                crate_path.version
+            )),
+
+            Some(release) => {
+                let analyzed_deps =
+                    analyze_dependencies(engine.clone(), release.deps.clone()).await?;
+
+                let crates = vec![(crate_path.name, analyzed_deps)].into_iter().collect();
```
Comment (Enet4) [inline src/engine/mod.rs:227]:
```suggestion
                let crates = vec![(crate_path.name, analyzed_deps)];
```

**Review by Enet4 — COMMENTED** [review]
Only one odd thing caught my eye in the diffs. Other than that, I'm OK with merging.

File: src/engine/mod.rs (hunk 2 above)
Comment (robjtede) [reply src/engine/mod.rs]: I imagine that clippy will find plenty more things like this soon

**Review by Enet4 — APPROVED** [review]

**Pull request merged**